### PR TITLE
Added files should be in the working directories.

### LINKED ISSUE #292: Copy added Spark files to the working directory of the driver and executors
Spark's contract is that files sent to a Spark job via `spark.files` are in the working directory of the driver and executor processes. The first version of the driver took the files sent to the submission server and wrote them into the working directory before starting the driver JVM. Submission v2 does not do this.

There's a few ways to accomplish this in the new architecture. The tricky part is that the init-container should be unaware in the strictest sense of what the working directory of the driver is. For example, users who change the working directory of the driver with custom Docker images will inadvertently make their driver have a different working directory from the init-container. Thus the design that is chosen should be resilient to this.

### PR DESCRIPTION
**Pull request by mccheah**
Closes #292.

The implementation here unfortunately requires more logic in the Dockerfile's command, which further increases the surface area of the API between the Docker image and the submission client. Nevertheless this seems like the best approach among the sub-optimal alternatives. For example, we could have made the init-container copy the files into its working directory, but this introduces an opaque contract where the init-container has to have the same working directory as the driver; since the working directories are defined by the Docker images, custom drivers may not be able to satisfy this guarantee.

The approach given here at least ensures that if files are added by any means to the given directory, that those files are at least consistent in being added to its own working directory, regardless of the state of the component that mounted those files in the first place.

### CODE REVIEW

File: resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile [hunk 1]
```diff
@@ -40,4 +40,5 @@ CMD SPARK_CLASSPATH="${SPARK_HOME}/jars/*" && \
     if ! [ -z ${SPARK_MOUNTED_CLASSPATH+x} ]; then SPARK_CLASSPATH="$SPARK_MOUNTED_CLASSPATH:$SPARK_CLASSPATH"; fi && \
     if ! [ -z ${SPARK_SUBMIT_EXTRA_CLASSPATH+x} ]; then SPARK_CLASSPATH="$SPARK_SUBMIT_EXTRA_CLASSPATH:$SPARK_CLASSPATH"; fi && \
     if ! [ -z ${SPARK_EXTRA_CLASSPATH+x} ]; then SPARK_CLASSPATH="$SPARK_EXTRA_CLASSPATH:$SPARK_CLASSPATH"; fi && \
+    if ! [ -z ${SPARK_MOUNTED_FILES_DIR} ]; then cp -R "$SPARK_MOUNTED_FILES_DIR/." .; fi && \
```
Comment (ash211) [inline resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile:43]: maybe can use `ln` instead of `cp` to reduce disk churn?

**Review by ash211 — COMMENTED** [review]
Looks good.  It seems like we should be able to have the init container write files directly into the working directory of the driver, as configured via an environment variable on the init container.  The contract then is that the init container must download the jars and place them in the dir specified by the envvar.

Do you feel that's an opaque contract?

**Comment by mccheah** [comment]
The problem is the submission client doesn't know what the working directory of the driver is. In our current model we've set the working directory in the Dockerfile and the submission client does not know many details of it.

File: resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile (hunk 1 above)
Comment (mccheah) [reply resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile:43]: We don't want to symbolic link the directory itself, but we can create a link to each file in the directory?

Comment (ash211) [reply resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile:43]: Right, could do some unix-foo to make the files appear in the other directory.  Something like:

`if ! [ -z ${SPARK_MOUNTED_FILES_DIR} ]; then find "$SPARK_MOUNTED_FILES_DIR/" -type f -exec ln -s {} . \; ; fi && \`
 
This is something we can do after this PR though -- would you rather revisit later if profiling reveals the copy is slow?  It should only affect large files.  If we push, I can file an issue to remind us to come back to this

Comment (mccheah) [reply resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile:43]: Yeah - `spark.files` is meant to be for small things, usually.

**Comment by ash211** [comment]
Ah, I thought `/opt/spark` was available somewhere outside the docker container but guess not.

Given that, I think this is the right approach.

**Review by ash211 — APPROVED** [review]

File: resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile (hunk 1 above)
Comment (ash211) [reply resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile:43]: https://github.com/apache-spark-on-k8s/spark/issues/295

**Comment by ash211** [comment]
Planning to merge when integration test passes

**Pull request merged**